src/compiler/collections.ts

```diff
@@ -0,0 +1,624 @@
+// NumberMap, StringMap, and StringSet shims
+/* @internal */
+namespace ts {
+    // The global Map object. This may not be available, so we must test for it.
+    declare const Map: NumberMapConstructor & StringMapConstructor | undefined;
+    const usingNativeMaps = typeof Map !== "undefined";
+    // tslint:disable-next-line:no-in-operator
+    const fullyFeaturedMaps = usingNativeMaps && "keys" in Map.prototype && "values" in Map.prototype && "entries" in Map.prototype;
+
+    /** Extra Map methods that may not be available, so we must provide fallbacks. */
+    interface ES6Map<K, V> extends Map<K, V> {
+        keys(): Iterator<K>;
+        values(): Iterator<V>;
+        entries(): Iterator<[K, V]>;
+    }
+
+    /** Simplified ES6 Iterator interface. */
+    interface Iterator<T> {
+        next(): { value: T, done: false } | { value: never, done: true };
+    }
+
+    export interface NumberMapConstructor {
+        /**
+         * Creates a new Map with number keys.
+         * If `pairs` is provided, each [key, value] pair will be added to the map.
+         */
+        new<K extends number, V>(pairs?: [K, V][]): Map<K, V>;
+    }
+
+    /**
+     * In runtimes without Maps, this is implemented using a sparse array.
+     * This is generic over the key type because it is usually an enum.
+     */
+    export const NumberMap: NumberMapConstructor = usingNativeMaps ? Map : class ShimNumberMap<K extends number, V> implements Map<K, V> {

```

Comment:
It ends up taking 20% longer. 🤕 

##### Monaco - node (v6.7.0, x86)

| Project    |          Baseline |           Current |             Delta |  Best |  Worst |
| ---------- | ----------------- | ----------------- | ----------------- | ----- | ------ |
| Memory used | 203,967k (±  0.01%) | 224,241k (±  0.01%) | +20,274k (+  9.94%) | 224,220k | 224,263k |
| Parse Time  |    2.27s (±  0.39%) |    2.39s (±  3.65%) |   +0.12s (+  5.30%) |    2.32s |    2.51s |
| Bind Time   |    0.68s (±  1.17%) |    0.83s (±  1.52%) |   +0.15s (+ 21.41%) |    0.81s |    0.84s |
| Check Time  |    3.86s (±  2.12%) |    4.72s (±  3.44%) |   +0.86s (+ 22.15%) |    4.55s |    4.89s |
| Emit Time   |    1.62s (±  2.07%) |    2.30s (± 14.05%) |   +0.68s (+ 41.80%) |    2.04s |    2.72s |
| Total Time  |    8.43s (±  1.26%) |   10.23s (±  3.36%) |   +1.80s (+ 21.34%) |    9.78s |   10.47s |

##### TFS - node (v6.7.0, x86)

| Project    |          Baseline |           Current |             Delta |  Best |  Worst |
| ---------- | ----------------- | ----------------- | ----------------- | ----- | ------ |
| Memory used | 175,828k (±  0.01%) | 188,541k (±  0.01%) | +12,713k (+  7.23%) | 188,529k | 188,551k |
| Parse Time  |    1.40s (±  0.75%) |    1.51s (±  4.24%) |   +0.11s (+  7.80%) |    1.46s |    1.60s |
| Bind Time   |    0.57s (±  1.02%) |    0.68s (±  2.26%) |   +0.12s (+ 20.63%) |    0.67s |    0.70s |
| Check Time  |    3.00s (±  0.53%) |    3.60s (±  3.55%) |   +0.60s (+ 19.88%) |    3.41s |    3.69s |
| Emit Time   |    1.37s (±  1.09%) |    1.71s (±  5.74%) |   +0.34s (+ 25.18%) |    1.66s |    1.86s |
| Total Time  |    6.33s (±  0.28%) |    7.50s (±  2.73%) |   +1.16s (+ 18.38%) |    7.24s |    7.71s |

Here's the code I tested with (with equivalents for NumberMap and StringSet):

```ts
export const createStringMap: <T>(pairs?: [string, T][]) => Map<string, T> = usingNativeMaps
    ? <T>(pairs?: [string, T][]) => new Map(pairs)
    : <T>(pairs?: [string, T][]) => {
        let data = createDictionaryModeObject<T>();
        if (pairs) {
            for (const [key, value] of pairs) {
                data[key] = value;
            }
        }
        return {
            clear(): void {
                data = createDictionaryModeObject<T>();
            },
            delete(key: string): void {
                delete data[key];
            },
            get(key: string): T {
                return data[key];
            },
            has(key: string): boolean {
                // tslint:disable-next-line:no-in-operator
                return key in data;
            },
            set(key: string, value: T): void {
                data[key] = value;
            },
            forEach(action: (value: T, key: string) => void): void {
                for (const key in data) {
                    action(data[key], key);
                }
            }
        }
    }
```

I bet this is allocating a different closure for each method, wheras a class just allocates space for `data` and shares a single prototype among all instances.